Merge Git for Windows v2.32.0.2

dscho opened the pull request:
This brings a couple bug fixes to this here fork:

- `git subtree` works again
- the built-in file system watcher now works in worktrees whose path contains non-ASCII characters
- client-side certificates can be used via Secure Channel again
- scrolling the output of `git log` no longer messes up (duplicating lines and some such)

dscho committed:
subtree: fix the GIT_EXEC_PATH sanity check to work on Windows

In 22d550749361 (subtree: don't fuss with PATH, 2021-04-27), `git
subtree` was broken thoroughly on Windows.

The reason is that it assumes Unix semantics, where `PATH` is
colon-separated, and it assumes that `$GIT_EXEC_PATH:` is a verbatim
prefix of `$PATH`. Neither are true, the latter in particular because
`GIT_EXEC_PATH` is a Windows-style path, while `PATH` is a Unix-style
path list.

Let's make extra certain that `$GIT_EXEC_PATH` and the first component
of `$PATH` refer to different entities before erroring out.

We do that by using the `test <path1> -ef <path2>` command that verifies
that the inode of `<path1>` and of `<path2>` is the same.

Sadly, this construct is non-portable, according to
https://pubs.opengroup.org/onlinepubs/009695399/utilities/test.html.
However, it does not matter in practice because we still first look
whether `$GIT_EXEC_PREFIX` is string-identical to the first component of
`$PATH`. This will give us the expected result everywhere but in Git for
Windows, and Git for Windows' own Bash _does_ handle the `-ef` operator.

Just in case that we _do_ need to show the error message _and_ are
running in a shell that lacks support for `-ef`, we simply suppress the
error output for that part.

This fixes https://github.com/git-for-windows/git/issues/3260

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
subtree: fix assumption about the directory separator

On Windows, both forward and backslash are valid separators. In
22d550749361 (subtree: don't fuss with PATH, 2021-04-27), however, we
added code that assumes that it can only be the forward slash.

Let's fix that.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Merge pull request #3263 from dscho/fix-built-in-fsmonitor

FSMonitor fixes

dscho committed:
pager: avoid setting COLUMNS when we're guessing its value

We query `TIOCGWINSZ` in Git to determine the correct value for
`COLUMNS`, and then set that environment variable.

If `TIOCGWINSZ` is not available, we fall back to the hard-coded value
80 _and still_ set the environment variable.

On Windows this is a problem. The reason is that Git for
Windows uses a version of `less` that relies on the MSYS2 runtime to
interact with the pseudo terminal (typically inside a MinTTY window,
which is also aware of the MSYS2 runtime). Both MinTTY and `less.exe`
interact with that pseudo terminal via `ioctl()` calls (which the MSYS2
runtime emulates even if there is no such thing on Windows).
Since https://github.com/gwsw/less/commit/bb0ee4e76c2, `less` prefers
the `COLUMNS` variable over asking ncurses itself.

But `git.exe` itself is _not_ aware of the MSYS2 runtime, or for that
matter of that pseudo terminal, and has no way to call `ioctl()` or
`TIOCGWINSZ`.

Therefore, `git.exe` will fall back to hard-coding 80 columns, no matter
what the actual terminal size is.

But `less.exe` is totally able to interact with the MSYS2 runtime and
would not actually require Git's help (which actually makes things
worse here). So let's not override `COLUMNS` on Windows.

Let's just not set `COLUMNS` unless we managed to query the actual value
from the terminal.

This fixes https://github.com/git-for-windows/git/issues/3235

Co-authored-by: Junio C Hamano <gitster@pobox.com>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! Add a GitHub workflow to monitor component updates

Let's watch out for new libgcrypt versions, too (this component is a
pretty important dependency of gnupg).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! Add a GitHub workflow to monitor component updates

Since we're now watching out for new libgcrypt versions, let's also
watch out for its dependency libgpg-error.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Merge pull request #3275 from dscho/monitor-libgcrypt

Monitor libgpg-error and libgcrypt versions

jeffhostetler committed:
fixup! fsmonitor: introduce `core.useBuiltinFSMonitor` to call the daemon via IPC

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

dscho committed:
Merge pull request #3287 from dscho/redo-fsmonitor-null-last-update

Redo fix for FSMonitor

pascalmuller committed:
http: optionally send SSL client certificate

This adds support for a new http.sslAutoClientCert config value.

In cURL 7.77 or later the schannel backend does not automatically send
client certificates from the Windows Certificate Store anymore.

This config value is only used if http.sslBackend is set to "schannel",
and can be used to opt in to the old behavior and force cURL to send
client certificates.

This fixes https://github.com/git-for-windows/git/issues/3292

Signed-off-by: Pascal Muller <pascalmuller@gmail.com>

dscho committed:
Merge pull request #3293 from pascalmuller/http-support-automatically-sending-client-certificate

http: Add support for enabling automatic sending of SSL client certificate

dscho committed:
fixup! README.md: Add a Windows-specific preamble

Add a section instructing developers how to build, and how to test the
just-built executables in-place.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Merge pull request #3297 from dscho/docs-about-testing-git-in-place

Document how to build (and test) Git for Windows

dscho committed:
fixup! fsmonitor-fs-listen-win32: implement FSMonitor backend on Windows

Let's keep avoiding the `*A()` family of Win32 API functions because
they are susceptible to incoherent encoding problems. In Git for
Windows, we always assume paths to be UTF-8 encoded. Let's use the
dedicated helper to convert such a path to the wide character version,
and then use the `*W()` function instead.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! mingw: support long paths

Let's allow long paths in the FSMonitor daemon, too.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Merge pull request #3272 from dscho/fix-duplicated-lines-when-moving-in-pager

Fix duplicated lines when scrolling in the pager

dscho committed:
Merge pull request #3264 from dscho/fix-subtree-on-windows

Fix subtree on Windows

dscho committed:
Merge pull request #3309 from dscho/fsmonitor-and-unicode-paths

fsmonitor: handle non-ASCII characters in the git_dir/worktree paths

dscho committed:
Merge tag 'v2.32.0.windows.2' into vfs-v2.32.0

This merges Git for Windows v2.32.0(2). There were only merge conflicts
in the `README.md` file, which sprouted some build instructions on the
Git for Windows side, and was revamped to describe Scalar on the
`microsoft/git` side (and why the fork exists in the first place.)

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

derrickstolee reviewed (approved):
I suppose we could cut a new `.0.5` (draft) release in anticipation of the upcoming VFS for Git release, since we have a large number of users waiting to upgrade to 2.32.0 there. Unless these are causing serious issues for our `microsoft/git` users, I wouldn't necessarily push a new update for Scalar users right now.

jeffhostetler commented:
Could we replace your commit:
6f50cde5df (fixup! fsmonitor-fs-listen-win32: implement FSMonitor backend on Windows, 2021-07-05)

with these from my https://github.com/jeffhostetler/git/commits/builtin-fsmonitor
c90e8c65db (fixup! fsmonitor-fs-listen-win32: implement FSMonitor backend on Windows, 2021-07-05)
d989fc7cbd (t7527: test builtin FSMonitor watching repos with unicode paths, 2021-07-12)

I had a problem cherry-picking your commit into upstream since the _long_ version isn't present yet and it looked like using the _long_ version should go into your next commit with the updated array declaration.
And I added a unit test to confirm that it works.

dscho commented:
> I had a problem cherry-picking your commit into upstream since the _long_ version isn't present yet and it looked like using the _long_ version should go into your next commit with the updated array declaration.

Yes, I noticed that problem myself, when I fixed the `shears/seen` rebase... Sorry about that!

> Could we replace your commit:
> [6f50cde](https://github.com/microsoft/git/commit/6f50cde5dff787cc5ba0eae7bd225550496a3708) (fixup! fsmonitor-fs-listen-win32: implement FSMonitor backend on Windows, 2021-07-05)
> 
> with these from my https://github.com/jeffhostetler/git/commits/builtin-fsmonitor
> [c90e8c6](https://github.com/microsoft/git/commit/c90e8c65db2eb9940d1c1f9a5bbecb417760af96) (fixup! fsmonitor-fs-listen-win32: implement FSMonitor backend on Windows, 2021-07-05)
> [d989fc7](https://github.com/microsoft/git/commit/d989fc7cbd9c52c6a1529a2de4519fcd0c771063) (t7527: test builtin FSMonitor watching repos with unicode paths, 2021-07-12)

Will do that in a moment.

dscho committed:
fixup! fsmonitor-fs-listen-win32: implement FSMonitor backend on Windows

Jeff Hostetler pointed out that this is incorrect (due to its use of
`MAX_LONG_PATH`). Let's drop it in preparation for cherry-picking a
better version.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fsmonitor-fs-listen-win32: implement FSMonitor backend on Windows

Let's keep avoiding the `*A()` family of Win32 API functions because
they are susceptible to incoherent encoding problems. In Git for
Windows, we always assume paths to be UTF-8 encoded. Let's use the
dedicated helper to convert such a path to the wide character version,
and then use the `*W()` function instead.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
t7527: test builtin FSMonitor watching repos with unicode paths

Create some test repos with UTF8 pathnames and verify that
the builtin FSMonitor can watch them.  This test is mainly
for Windows where we need to avoid `*A()` routines.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
@jeffhostetler hmm...:

> ./test-lib.sh: line 903: stop_daemon_delete_repo: command not found

Maybe we should skip the tip commit, then?

dscho commented:
@jeffhostetler also, while the new `fixup!` commit is the correct one to send upstream, the old one at least results in the correct post image when targeting Git for Windows (and, by extension, `microsoft/git`): the end result needs to handle long paths.

Maybe I should go back to what I had before?

jeffhostetler commented:
I'm not sure I understand the subtleties here of which version of the *A to *W fix you should take.
The one without the \_long\_ needs to go upstream.  I was just trying to avoid you having a merge
conflict later when/if GFW inherits FSMonitor from upstream.  So take whichever version of the fixup
you need here and we can deal with it later.

WRT the error in the test, I'm not sure what's going on here.  There might be a quoting issue or
something, but we can figure that out later.  So yeah, just drop it for now.

jeffhostetler commented:
I have to wonder if we need an `export` on the function to have it be seen inside the layers of subshells created by the for loop....

dscho commented:
> I'm not sure I understand the subtleties here of which version of the *A to *W fix you should take.
> The one without the _long_ needs to go upstream.

Right, because upstream does not have support for `core.longPaths` yet, but microsoft/git does.

> I was just trying to avoid you having a merge conflict later when/if GFW inherits FSMonitor from upstream. So take whichever version of the fixup you need here and we can deal with it later.

That's very considerate of you! The merge conflict will happen in git-for-windows/git, though, and it will hit the person responsible for it: me ;-)

> WRT the error in the test, I'm not sure what's going on here. There might be a quoting issue or
> something, but we can figure that out later. So yeah, just drop it for now.

I think it is a function that is missing in microsoft/git because we're still running with v2, not with v3, of the FSMonitor patch seres.

> I have to wonder if we need an `export` on the function to have it be seen inside the layers of subshells created by the for loop....

Doesn't it work in your PR? If so, then you probably introduced that function in v3.

dscho commented:
After discussing this in the team, we decided that it is not necessary to integrate this, we'll just wait for the next Git for Windows version to get all of this.
